### PR TITLE
AP_InertialSensor: support logging of batches of raw IMU data

### PR DESCRIPTION
**Pull request by peterbarker**
Full-on logging of raw IMU data is too expensive for people to leave on all the time, in terms of CPU time, SD card and SD bandwidth.

This PR implements an alternative to the current raw-IMU logging.  Sensor data is taken at full rate into an allocated buffer (in the relevant sensor thread).  This data is then streamed out to the SD card in the main thread at a constant rate (1kB/s).  Each IMU sensor is sampled in turn.

The pymavlink tools directory contains `mavfft_ibs.py` (https://github.com/ArduPilot/pymavlink/blob/master/tools/mavfft_isb.py), a tool to extract the samples from a log and create a frequency analysis plot from that data.

A n Issue has been created for Mission Planner (https://github.com/ArduPilot/MissionPlanner/issues/1647).

### CODE REVIEW

File: libraries/AP_InertialSensor/BatchSampler.cpp [hunk 1]
```diff
@@ -0,0 +1,198 @@
...
+{
+    const uint8_t old_instance = instance;
+    do {
+        instance++;
+        switch(type) {
+        case IMU_SENSOR_TYPE_ACCEL:
+            if (instance >= _imu._accel_count) {
+                type = IMU_SENSOR_TYPE_GYRO;
+                instance = 0;
+                multiplier = multiplier_accel;
```
Comment (tridge) [inline libraries/AP_InertialSensor/BatchSampler.cpp]: I think this should be multiplier_gyro

File: libraries/AP_InertialSensor/BatchSampler.cpp [hunk 2]
```diff
@@ -0,0 +1,198 @@
...
+            }
+            break;
+        case IMU_SENSOR_TYPE_GYRO:
+            if (instance >= _imu._gyro_count) {
+                type = IMU_SENSOR_TYPE_ACCEL;
+                instance = 0;
+                multiplier = multiplier_gyro;
```
Comment (tridge) [inline libraries/AP_InertialSensor/BatchSampler.cpp]: this should be multipler_accel
I think this should the code should be restructured a bit to make it clearer. Maybe a first switch to change the type, then a 2nd switch to setup the new multuplier

File: libraries/AP_InertialSensor/BatchSampler.cpp [hunk 3]
```diff
@@ -0,0 +1,198 @@
...
+            }
+            break;
+        }
+        if (_sensor_mask & (1U<<instance)) {
+            break;
+        }
+    } while (instance != old_instance);
```
Comment (tridge) [inline libraries/AP_InertialSensor/BatchSampler.cpp]: don't we risk looping forever here?

File: libraries/AP_InertialSensor/BatchSampler.cpp [hunk 4]
```diff
@@ -0,0 +1,198 @@
...
+    }
+    if (_sensor_mask == 0) {
+        return;
+    }
+    if (data_write_offset - data_read_offset < samples_per_msg) {
+        // insuffucient data to pack a packet
+        return;
+    }
+    const uint64_t now = AP_HAL::micros64();
+    if (last_sent_ms - now < push_interval_ms) {
```
Comment (tridge) [inline libraries/AP_InertialSensor/BatchSampler.cpp]: it is not safe to subtract uint64_t and uint32_t - both must be the same width

Comment (tridge) [reply libraries/AP_InertialSensor/BatchSampler.cpp]: also need to be the same units! don't compare milliseconds and microseconds

File: libraries/AP_InertialSensor/BatchSampler.cpp (hunk 2 above)
Comment (peterbarker) [reply libraries/AP_InertialSensor/BatchSampler.cpp]: Will fix.  I've just factored out the setting of multiplier.  It doesn't help much.

File: libraries/AP_InertialSensor/BatchSampler.cpp (hunk 4 above)
Comment (peterbarker) [reply libraries/AP_InertialSensor/BatchSampler.cpp]: Will fix.

File: libraries/AP_InertialSensor/BatchSampler.cpp (hunk 3 above)
Comment (peterbarker) [reply libraries/AP_InertialSensor/BatchSampler.cpp]: Only if the number of gyros or accels change, I think.

I've changed this to a pair of for loops, and used the assumption (as discussed) that there will always be an equal number of gyros and accels.

**Comment by peterbarker** [comment]
@OXINARF why the "NeedToChangeGCS" tag?  Certainly there could be benefits to changing them, but it should be a prereq for merging this.

File: libraries/AP_InertialSensor/AP_InertialSensor.h [hunk 5]
```diff
@@ -268,6 +273,74 @@ class AP_InertialSensor : AP_AccelCal_Client
     // return time in microseconds of last update() call
     uint32_t get_last_update_usec(void) const { return _last_update_usec; }
 
+    enum IMU_SENSOR_TYPE {
+        IMU_SENSOR_TYPE_ACCEL,
```
Comment (tridge) [inline libraries/AP_InertialSensor/AP_InertialSensor.h]: minor point, given this forms part of the log data (so part of a protocol), I think giving the specific values with =0 and =1 is worthwhile in the enum

Comment (peterbarker) [reply libraries/AP_InertialSensor/AP_InertialSensor.h]: Done.

**Comment by peterbarker** [comment]
@tridge I've updated the expected time taken for the scheduled `ins.periodic()` based on `perf`'s output.

**Comment by OXINARF** [comment]
> why the "NeedToChangeGCS" tag? Certainly there could be benefits to changing them, but it should be a prereq for merging this.

@peterbarker What happens when a GCS tries to analyse a log that has an *a* format in some message? It can certainly try to go over the message, but that's a fallback, not the correct way to operate.

**Comment by peterbarker** [comment]
On Thu, 5 Oct 2017, Francisco Ferreira wrote:

>
>       why the "NeedToChangeGCS" tag? Certainly there could be benefits
>       to changing them, but it should be a prereq for merging this.
> 
> @peterbarker What happens when a GCS tries to analyse a log that has an a
> format in some message? It can certainly try to go over the message, but
> that's a fallback, not the correct way to operate.

Skipping over the message is the only practical way forward.  Thankfully 
we know the size of the message so that's easy enough.

We had this problem when we added the Q and q types - that was a lot more 
disruptive as we changes to using TimeUS with those :-)

**Comment by OXINARF** [comment]
> Skipping over the message is the only practical way forward.  Thankfully
we know the size of the message so that's easy enough.

If a GCS is changing code to skip over these specific messages then it's much easier to support the *a* format instead.
Just to clarify: I'm not saying that the PR should be blocked until all GCS have adapted, just saying that GCS need to be changed.

**Comment by peterbarker** [comment]
On Fri, 6 Oct 2017, Francisco Ferreira wrote:

>       Skipping over the message is the only practical way forward.
>       Thankfully
>       we know the size of the message so that's easy enough.
> 
> If a GCS is changing code to skip over these specific messages then it's
> much easier to support the a format instead.
> Just to clarify: I'm not saying that the PR should be blocked until all GCS
> have adapted, just saying that GCS need to be changed.

Now I'm with you.

You're almost certainly correct, at least in some cases.  We probably 
should have learnt when we added 'Q' and 'q' support to skip messages with 
fields we didn't understand, but I'm pretty sure the parsers I've worked 
with don't do that....

**Comment by peterbarker** [comment]
On Fri, 6 Oct 2017, Francisco Ferreira wrote:

> > @@ -106,6 +108,17 @@ class DataFlash_Class
>      void Log_Write_RFND(const RangeFinder &rangefinder);
>      void Log_Write_IMU(const AP_InertialSensor &ins);
>      void Log_Write_IMUDT(const AP_InertialSensor &ins, uint64_t time_us, ui
> nt8_t imu_mask);
> +    bool Log_Write_ISBH(uint16_t seqno,
> +                        AP_InertialSensor::IMU_SENSOR_TYPE sensor_type,
> +                        uint8_t instance,
> +                        uint16_t multiplier,
> +                        uint64_t sample_us,
> +                        float sample_rate_hz);
> +    bool Log_Write_ISBD(uint16_t isb_seqno,
> +                        uint16_t seqno,
> +                        const int16_t x[32],
> +                        const int16_t y[32],
> +                        const int16_t z[32]);
> 
> const? 😜

I *think* that's all correct?  Not const'ing stack stuff but const'ing the 
more complicated ones....

> In libraries/AP_Math/AP_Math.h:
> 
> > @@ -154,7 +154,7 @@ inline int32_t constrain_int32(const int32_t amt, cons
> t int32_t low, const int32
>  }
>
>  // degrees -> radians
> -static inline float radians(float deg)
> +static inline constexpr float radians(float deg)
> 
> It would be nice to put the same in degrees

Will do.

> ____________________________________________________________________________
> 
> In libraries/AP_InertialSensor/BatchSampler.cpp:
> 
> > +
> +
> +extern const AP_HAL::HAL& hal;
> +void AP_InertialSensor::BatchSampler::init()
> +{
> +    if (_sensor_mask == 0) {
> +        return;
> +    }
> +    if (_required_count <= 0) {
> +        return;
> +    }
> +
> +    _required_count -= _required_count % 32; // round down to nearest multi
> ple of 32
> +
> +    const uint32_t total_allocation = 3*_required_count*sizeof(uint16_t);
> +    gcs().send_text(MAV_SEVERITY_WARNING, "INS: alloc %u bytes for ISB (fre
> e=%u)", total_allocation, hal.util->available_memory());
> 
> Does this really need to be a warning? Even info looks too much, looks more
> like debug.

Changed to debug.

> ____________________________________________________________________________
> 
> In libraries/AP_InertialSensor/BatchSampler.cpp:
> 
> > +
> +    if (type == IMU_SENSOR_TYPE_ACCEL) {
> +        // we have logged accelerometers, now log gyros:
> +        type = IMU_SENSOR_TYPE_GYRO;
> +        multiplier = multiplier_gyro;
> +        return;
> +    }
> +
> +    // log for accel sensor:
> +    type = IMU_SENSOR_TYPE_ACCEL;
> +    multiplier = multiplier_accel;
> +
> +    // move to next IMU backend:
> +
> +    // we assume the number of gyros and accels is the same:
> +    const uint8_t max_count = MIN(_imu._accel_count, _imu._gyro_count);
> 
> This seems strange:
>
>  *  the comment says we assume the number of gyros and accels is the same
>     but then we are taking a minimum between them (if we assume they are the
>     same, just take one of them, no need for the comparison)

The comment's correct, but obviously needs some help.  My first 
implementation didn't assume the counts were equal, but handling that 
wasn't pretty.  The min() just stops us dying if the counts are not equal.

I've adjusted the comment to indicate why we're doing the min() in 
particular.

>  *  the variable is named max_count but it is taking the minimum number of
>     accels/gyros

Heh.  Changed it to simply "_count"

Thanks!

**Comment by peterbarker** [comment]
On Mon, 9 Oct 2017, Francisco Ferreira wrote:

> @OXINARF approved this pull request.
> 
> LGTM.
> 
> Do you have any plans to enable this by default in the future?

That's indeed the intent.  The IMU mask was primarily added at this point 
to give us a way of disabling the feature until we've got some "all good, 
no fires, floods or plagues of locusts were created when I enabled this 
option" reports.

It's possible that when we do enable it by default that we might exclude 
some known-bad-and-unusable IMUs based on board type.

File: libraries/AP_InertialSensor/BatchSampler.cpp [hunk 6]
```diff
@@ -0,0 +1,216 @@
...
+    }
+    if (_sensor_mask == 0) {
+        return;
+    }
+    if (data_write_offset - data_read_offset < samples_per_msg) {
+        // insuffucient data to pack a packet
+        return;
+    }
+    const uint32_t now = AP_HAL::millis();
+    if (last_sent_ms - now < push_interval_ms) {
```
Comment (tridge) [inline libraries/AP_InertialSensor/BatchSampler.cpp]: this should be now - last_sent_ms

File: libraries/AP_InertialSensor/BatchSampler.cpp [hunk 7]
```diff
@@ -0,0 +1,216 @@
...
+void AP_InertialSensor::BatchSampler::sample(uint8_t _instance, AP_InertialSensor::IMU_SENSOR_TYPE _type, uint64_t sample_us, const Vector3f &_sample)
+{
+    if (!should_log(_instance, _type)) {
+        return;
+    }
+    if (data_write_offset == 0) {
+        measurement_started_us = sample_us;
+    }
+
+    data_x[data_write_offset] = multiplier*_sample.x;
```
Comment (tridge) [inline libraries/AP_InertialSensor/BatchSampler.cpp]: is there some reason to not use a Vector3i ?

File: libraries/AP_InertialSensor/AP_InertialSensor_Backend.cpp [hunk 8]
```diff
@@ -200,8 +200,17 @@ void AP_InertialSensor_Backend::_notify_new_gyro_raw_sample(uint8_t instance,
...
+}
+
+void AP_InertialSensor_Backend::log_gyro_raw(uint8_t instance, const uint64_t sample_us, const Vector3f &gyro)
+{
+    DataFlash_Class *dataflash = DataFlash_Class::instance();
+    if (dataflash == nullptr) {
+        // should not have been called
+        return;
+    }
+    if (should_log_imu_raw()) {
```
Comment (tridge) [inline libraries/AP_InertialSensor/AP_InertialSensor_Backend.cpp]: can we remove the old log_GYRO and log_ACCEL now?

**Comment by tridge** [comment]
I think this is ready to go in

**Pull request merged**